### PR TITLE
Implement access wrapping

### PR DESCRIPTION
**Pull request by no2chem**
This PR introduces the accessWrap and deepAccessWrap fields
to accessor annotations, which instructs the annotationProcessor
to build a thin wrapper around the return value of the method,
forcing any function calls on that object to be synchronized
against the runtime as if were called against the parent
object.

The deepAccessWrap parameter instructs the annotationProcessor
to wrap any function with the given string.

In addition a new SMRConcurrentMap class has been introduced, which doesn't
throw ConcurrentModificationException during iteration of keySet, entrySet and 
values.

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/collections/SMRConcurrentMap.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        }
+        return curValue;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    @TransactionalMethod
+    public V computeIfAbsent(K key, Function<? super K, ? extends V> mappingFunction) {
```
Comment (goyalkapil) [inline runtime/src/main/java/org/corfudb/runtime/collections/SMRConcurrentMap.java]: There's no such method to override?

File: runtime/src/main/java/org/corfudb/runtime/collections/SMRConcurrentMap.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+        return v;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    @TransactionalMethod
+    public V computeIfPresent(K key, BiFunction<? super K, ? super V, ? extends V> remappingFunction) {
```
Comment (goyalkapil) [inline runtime/src/main/java/org/corfudb/runtime/collections/SMRConcurrentMap.java]: There's no such method to override?

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **6ec012a**.

*** 1.3% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 3.61% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #627 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/IwIgCmN7iTS4AqNC0CK4Wa4gAYSZqHrT)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **6ec012a**.

*** 19.5% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 15.1375% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #627 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/Yf3DuqhZwy62uNrF8Ck23gI74j4jH7vO)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **6ec012a**.

*** 6.61% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 20.2175% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #627 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/GKAOGyPZhRfe2SLNYKlsbfa6M1hARHLh)

File: runtime/src/main/java/org/corfudb/runtime/collections/SMRConcurrentMap.java (hunk 1 above)
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/collections/SMRConcurrentMap.java]: it's overriding a method of the map interface.

File: runtime/src/main/java/org/corfudb/runtime/collections/SMRConcurrentMap.java (hunk 2 above)
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/collections/SMRConcurrentMap.java]: same, overriding map interface

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **315f762**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #627 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/Uz6jDTJZ0b9BzxAFN564dlTNsjVjj12o)

**Pull request closed without merging**